### PR TITLE
Correctly perform the fast-path in Evarsolve.invert_definition.

### PR DESCRIPTION
**Pull request by ppedrot**
The previous code was computing O(n) aliasing data before checking for the fast path.

### CODE REVIEW

**Comment by ppedrot** [comment]
@coqbot bench

File: pretyping/evarsolve.ml [hunk 1]
```diff
@@ -1600,11 +1600,31 @@ exception MetaOccurInBodyInternal
...
+    let rec is_id_subst ctxt s =
+      match ctxt, SList.view s with
+      | (decl :: ctxt'), Some (c, s') ->
+        let id = get_id decl in
+        names := Id.Set.add id !names;
+        (match c with None -> true | Some c -> isVarId evd id c) && is_id_subst ctxt' s'
+      | [], None -> true
+      | _ -> false
+    in
+      is_id_subst sign argsv &&
```
Comment (gares) [inline pretyping/evarsolve.ml:1617]: Isn't your new representation of evar subst telling you this? here you are rediscovering it. I was expecting the check to be: is the SList morally made of `length ctx`  `None` nodes.

Comment (ppedrot) [reply pretyping/evarsolve.ml:1617]: Yes, but you need to compute the names anyways. I think we can try to be even more clever but this requires a more in-depth modification, e.g. storing more data in evar infos.

Comment (ppedrot) [reply pretyping/evarsolve.ml:1617]: A quick analysis of the calls indicate that there is virtually a 100% hit for the fast path when the substitution is the compact identity. This means that in any case as conjectured above, we would always recompute the set of names even if we get an immediate answer. Thus, first checking for trivial identities wouldn't change the profile.

**Comment by ppedrot** [comment]
I've seen worse bench.

**Comment by ppedrot** [comment]
CI failures are unrelated.

**Comment by gares** [comment]
Do you want me to merge it before/after/together https://github.com/coq/coq/pull/16713 ? (just to help you timing things, in case you need to rebase the other one)

**Comment by ppedrot** [comment]
Please go with this one first. The other PR doesn't look so efficient anyways (except for degenerate examples).

**Comment by gares** [comment]
@coqbot merge now

**Review by gares — APPROVED** [review]

**Comment by gares** [comment]
@coqbot merge now

**Pull request merged**